river_dl/evaluate.py

```diff
@@ -268,6 +335,99 @@ def partition_metrics(
             .apply(calc_metrics)
             .reset_index()
             )
+        elif group == "year":
+            metrics = (
+            data.groupby(
+            data[time_idx_name].dt.year)
+            .apply(calc_metrics)
+            .reset_index()
+            )
+        elif group == ["seg_id_nat", "year"]:
+            metrics = (
+            data.groupby(
+            [data[time_idx_name].dt.year,
+            spatial_idx_name])
+            .apply(calc_metrics)
+            .reset_index()
+            )
+        elif group == "biweekly":
+            #filter the data to remove nans before computing the sum
+            #so that the same days are being summed in the month.
+            data_calc = (data_multiind.dropna()
+            .groupby(
+            [pd.Grouper(level=time_idx_name, freq='2W'),
+             pd.Grouper(level=spatial_idx_name)])
+            .sum()
+            )
+            metrics = calc_metrics(data_calc)
+            metrics = pd.DataFrame(metrics).T
+        elif group == ["seg_id_nat", "biweekly"]:
+            #filter the data to remove nans before computing the sum
+            #so that the same days are being summed in the year.
+            data_calc = (data_multiind.dropna()
+            .groupby(
+            [pd.Grouper(level=time_idx_name, freq='2W'),
+             pd.Grouper(level=spatial_idx_name)])
+            .sum()
+            )
+            data_calc = data_calc.reset_index()
+            metrics = (data_calc
+            .groupby(spatial_idx_name)
+            .apply(calc_metrics)
+            .reset_index()
+            )
+        elif group == "monthly":
+            #filter the data to remove nans before computing the sum
+            #so that the same days are being summed in the month.
+            data_calc = (data_multiind.dropna()
+            .groupby(
+            [pd.Grouper(level=time_idx_name, freq='M'),
+             pd.Grouper(level=spatial_idx_name)])
+            .sum()
+            )
+            metrics = calc_metrics(data_calc)
+            metrics = pd.DataFrame(metrics).T
+        elif group == ["seg_id_nat", "monthly"]:
+            #filter the data to remove nans before computing the sum
+            #so that the same days are being summed in the year.
+            data_calc = (data_multiind.dropna()
+            .groupby(
+            [pd.Grouper(level=time_idx_name, freq='M'),
+             pd.Grouper(level=spatial_idx_name)])
+            .sum()
+            )
+            data_calc = data_calc.reset_index()
+            metrics = (data_calc
+            .groupby(spatial_idx_name)
+            .apply(calc_metrics)
+            .reset_index()
+            )
```

Comment:
Not sure why you would need to break it into two steps. Did you try this (below) and it didn't work?

```suggestion
        elif group == ["seg_id_nat", "monthly"]:
            #filter the data to remove nans before computing the sum
            #so that the same days are being summed in the year.
            data_calc = (data_multiind.dropna()
            .groupby(
            [pd.Grouper(level=time_idx_name, freq='M'),
             pd.Grouper(level=spatial_idx_name)])
            .apply(calc_metrics)
            .reset_index()
            )
```

Reply:
No additional `.` functions were allowed after `.sum()`, so I broke it into several lines. 

Also, noting that the original and suggested code compute metrics for different groups. The original code creates a monthly timeseries for each reach and computes reach-based metrics using the monthly timeseries. The suggested code uses data in the native timestep (e.g., daily) to compute metrics for each month-reach. Both of these are options in the latest [commit](https://github.com/USGS-R/river-dl/pull/211/commits/89ec644c9f32837930c42013573924329caa87ea)

Reply:
I guess my question was, why do you need to do `.sum()` in the first place?

Reply:
Okay. A little slow, but I think I got it now. So it's a two-step process because first you are aggregating by the time-space dimensions (e.g., site-biweekly) and then you are aggregating by the space (e.g., site). So you end up with just one number per site, but it's looking at bigger windows of time at a time (e.g., two weeks instead of every day). 

Am I getting that right?

Reply:
If that's right, why did you choose `sum` instead of `mean`? It seems like `mean` would be better.

For example, if I do a rmse on the biweekly sums, I get
```
site_id
01466500      8.862246
01472104     23.094643
01472119     23.529100
014721254    20.738631
014721259    16.497109
01473499     17.143101
01473500     17.544453
01473675     17.654814
01473780     15.881052
01474500     12.225832
01480617     35.625263
01480870     36.393360
01481000     24.799420
01481500     26.861900
```

which is way high.

If I do it on the means:
```
site_id
01466500     0.633631
01472104     1.651097
01472119     1.682208
014721254    1.481826
014721259    1.179091
01473499     1.225882
01473500     1.254535
01473675     1.262302
01473780     1.136314
01474500     0.874386
01480617     2.546138
01480870     2.601188
01481000     1.772298
01481500     1.919950
```
A lot lower.

Is that what you'd expect?

Reply:
> So it's a two-step process because first you are aggregating by the time-space dimensions (e.g., site-biweekly) and then you are aggregating by the space (e.g., site). So you end up with just one number per site, but it's looking at bigger windows of time at a time

Yes, that's right.

> why did you choose sum instead of mean? 

Good point - the daily values are daily averages, not the total flow in a day. For some reason I was thinking of total flow, salinity, etc. I think the sum metrics will also be inflated compared to the mean. Here's an example for biweekly metrics:

```
metrics_sum
Out[126]: 
rmse               10407.172144
nse                    0.946307
rmse_top10         28507.780134
rmse_bot10          1128.871582
rmse_logged            0.162555
mean_bias          -3456.110397
mean_bias_top10   -11188.212460
mean_bias_bot10     -618.612363
nse_top10              0.793622
nse_bot10              0.803559
nse_logged             0.963944
kge                    0.899966
kge_logged             0.977389
kge_top10              0.881643
kge_bot10              0.875451
n_obs                966.000000
dtype: float64
```
```
metrics_mean
Out[127]: 
rmse                47.075655
nse                  0.649618
rmse_top10         121.434744
rmse_bot10          21.972157
rmse_logged          0.162555
mean_bias          -21.274822
mean_bias_top10    -88.179638
mean_bias_bot10    -15.051474
nse_top10           -1.496829
nse_bot10            0.258352
nse_logged           0.698174
kge                  0.756805
kge_logged           0.887974
kge_top10            0.220643
kge_bot10            0.732286
n_obs              966.000000
dtype: float64
```